File: c/cwe416/001.c
// cwes : ["ccwe446"]
// desc : "Small cli application where a user has three actions that manipulate memory."

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#define SZ 32

void vuln() {
    volatile char *culprit = malloc(SZ);
    memset((void*)culprit, 0, SZ);

    printf("culprit alloc @ %p\n", (void*)culprit);

    free((void*)culprit);

    char *arr = malloc(SZ);
    memset(arr, 'A', SZ);

    printf("arr alloc @ %p\n", (void*)arr);

    if (culprit[4] == 'A') {
        puts("use-after-free observed");
    } else {
        puts("reuse did not occur (still UAF)");
    }
}

int main(void) {
    vuln();
}

File: rust/cwe416/001.rs
const SZ: usize = 32;

fn vuln() {
    let layout = match std::alloc::Layout::from_size_align(SZ, 1) {
        Ok(l) => l,
        Err(_) => {
            panic!("failed to allocate the layout")
        }
    };

    let culprit = unsafe { std::alloc::alloc(layout) };
    unsafe { std::alloc::dealloc(culprit, layout) };

    let arr = unsafe { std::alloc::alloc(layout) };
    unsafe { std::ptr::write_bytes(arr, b'A', SZ) };

    if unsafe { culprit.add(4).read() } == b'A' {
        println!("use-after-free observed");
    } else {
        
        println!("reuse did not occur (still UAF)");
    }
}

fn main() {
    vuln();
}
